lib/miq_automation_engine/engine/miq_ae_method_service/miq_ae_service_model_base.rb

```diff
@@ -159,12 +159,13 @@ def self.expose(*args)
       options = args.last.kind_of?(Hash) ? args.pop : {}
       raise ArgumentError, "cannot have :method option if there is more than one method name specified" if options.key?(:method) && args.length != 1
 
+      user = User.current_user
       args.each do |method_name|
         next if method_name.to_sym == :id
         self.association = method_name if options[:association]
         define_method(method_name) do |*params|
           method = options[:method] || method_name
-          ret = object_send(method, *params)
+          ret = User.with_user(user) { object_send(method, *params) }
```

Comment:
The `user` variable is set outside of the `define_method` block and therefore will be capture when `expose` is run, which is at service model creation time.  The `user` captured at that time would always be used regardless of the user context when the method is run on an instance of the class.

Here is a simple example showing the results:
```ruby
class ABC
  def self.expose(method_name)
    user = $current_user
    define_method(method_name) do
      puts "User: #{user}"
    end
  end
end

$current_user = "Initial Name"
ABC.expose :test_method

ABC.new.test_method
# => User: Initial Name

$current_user = "Updated Name"
ABC.new.test_method
# => User: Initial Name
```

I believe what you are really trying to do is to run in the user context when the instance method is run, right?  For that you would need to determine the `user` inside of the `define_method` block.

Reply:
@gmcculloug 
Its the same user, the user is not changing under the covers. Instead the problem is that Drb runs requests coming from Automate Methods in different threads and those threads don't have the Thread Local Storage set for current user. This closure approach of wrapping the user before the DRb method runs will work. Once Automate starts it runs with the given user context we don't allow them to change it.

Reply:
My point is the class is going to get created at some point for a process and once it is created it will remain until the process ends.  A request from a different user that runs in the same process would be run in the context of the user when the class was created.  It is very possible the class would be created in a system context and callers to the method defined in `expose` would always have a `nil` user context.

Let me know if you want to discuss.

Reply:
@gmcculloug I was under the impression that the classes are going to get loaded for every Automate Request coming in since the service models are going to call the expose method.
If that is not true then we should just fetch the workspace from the DRb service front like its done here

https://github.com/ManageIQ/manageiq-automation_engine/blob/7b39eb1d8a4dba982bd33781a77bab8d728e5bf9/lib/miq_automation_engine/engine/miq_ae_method_service/miq_ae_service_rbac.rb#L39

The expose method should fetch the workspace and the workspace has a ae_user object that can be used with a with_user

Reply:
The `expose` method is called once during the class creation [1] and the method defined by `define_method` would be created at that time.  When the user calls the dynamically defined method, for example `ext_management_system` from the link below, the `user` value would already be set from creation time.

[1] https://github.com/ManageIQ/manageiq-automation_engine/blob/master/lib/miq_automation_engine/service_models/miq_ae_service_vm_or_template.rb#L14